README.md
hut
===
[![Build Status](https://travis-ci.org/okorz001/hut.svg?branch=master)](https://travis-ci.org/okorz001/hut)

A minimal framework for isomorphic apps with react and redux.

Usage
-----

This package contains both server and client code. Both sides expose a simple
object-oriented interface.

```js
import HutServer from 'hut/lib/server'

// Create the app.
const opts = {}
const app = new HutServer(opts)
// Start serving requests asynchronously.
app.run(() => console.log('running'))

// You can also stop the server asynchronously.
app.stop(() => console.log('stopped'))
```

```js
import HutClient from 'hut/lib/client'

// Create the app.
const opts = {}
const app = new HutClient(opts)
// Start rendering components synchronously.
app.run()
```

Options
-------

Some options are shared between both sides, while others are specific to each
environment. For shared options, it is recommended to use a single definition
and use `Object.assign` to derive server and client options.

```js
const commonOpts = {foo: 'x'}
const serverOpts = Object.assign({}, commonOpts, {bar: 'y'})
const clientOpts = Object.assign({}, commonOpts, {baz: 'z'})
```

### Common

#### `mountPoint: string`
default: `'app'`

The id of the HTML element where the application will be rendered and mounted.

#### `stateVariable: string`
default: `'initialState'`

The name of the global variable where the redux store state will be serialized.
The state that is used to render the page on the server is serialized and sent
to the client. The client uses this initial state to seed the store in the
browser.

#### `routes: react-router Routes`
default: `[]`

This route or array of routes is passed directly to the react-router library
as-is. The default value is not very useful as it will not match any paths.

[See react-router documentation for more details about configuring routes.](https://github.com/rackt/react-router/blob/latest/docs/guides/basics/RouteConfiguration.md)

#### `reducer: redux reducer function`
default: `state => return state || {}`

This reducer is passed directly to the redux library as-is. The default value
is not very useful as it ignores every action, creating a stateless app.

[See redux documentation for more details about reducers.](http://rackt.org/redux/docs/basics/Reducers.html)

#### `reduxMiddleware: array of redux middleware`
default: `[]`

An array of middleware that is applied directly to the redux createStore as-is,
and in the specified order.

[See redux documentation for more details about middleware.](http://rackt.org/redux/docs/advanced/Middleware.html)

### Server

#### `port: integer`
default: `process.env.PORT || 3000`

The port to listen on.

#### `middleware: array of koa middleware`
default: `[]`

An array of middleware that is applied directly to the koa server as-is,
and in the specified order.

[See koa documentation for more details about middleware.](https://github.com/koajs/koa/blob/master/docs/guide.md#writing-middleware)

#### `getTitle: function`
default: `() => ''`

A function that is called to generate the page title during rendering.
The default value is not very useful as it creates an empty title.

#### `scripts: array of strings`
default: `[]`

An array of URLs that are inserted as script tags into the page header.
The scripts are inserted with a defer tag so they will not block
rendering.

*If the script is a relative URL, you must also register middleware to
actually serve it.*

#### `styles: array of strings`
default: `[]`

An array of URLs that are inserted as link tags into the page header.

*If the style is a relative URL, you must also register middleware to
actually serve it.*

### Client

#### `createHistory: function`
default: `history.createBrowserHistory`

A function that is called to control navigation and save history for
the app. The return value is passed directly to react-router as-is.
The default value creates a history that uses the HTML5 history API.

[See history documentation for more details about histories.](https://github.com/rackt/history)


CHANGELOG.md
## [0.1.1](https://github.com/okorz001/hut/compare/v0.1.0...v0.1.1)

* Remove undocumented empty body check. Koa middleware that renders a body
  should not yield next.
* Support async onEnter Route hooks. Rendering will block until all onEnter
  hooks are completed.


test/server.js
import {expect} from 'chai'
import {load} from 'cheerio'
import React from 'react'
import {Route, Redirect} from 'react-router'
import request from 'request'

import SmushServer from '../src/server.js'

// Conveience method for verifying the number of elements matching a selector.
// Adds the selector name to the assertion failure for clarity.
function expectSelector($, selector, length) {
  expect($(selector), selector).to.have.length(length)
}

describe('SmushServer', function () {
  const mountPoint = 'mount'
  const initialState = 'zzz'
  const scripts = ['/a.js', '/b.js']
  const styles = ['/a.css', '/b.css']
  const port = 8080
  const state = 42

  function getTitle() {
    return 'Sweet Title'
  }

  function get(path, cb) {
    request({
      url: `http://localhost:${port}${path}`,
      // So we can see any 30x status codes.
      followRedirect: false,
    }, (err, resp, body) => {
      // Only call load if body is defined.
      cb(err, resp, body && load(body))
    })
  }

  function Foo() {
    return <div className="foo"></div>
  }

  function Bar() {
    return <div className="bar"></div>
  }

  function Params(props) {
    return <div data-a={props.params.a} data-b={props.location.query.b}></div>
  }

  function onEnterWait(newState, replaceState, done) {
    setTimeout(done, 1000)
  }

  const routes = [
    <Route path="/foo" component={Foo} />,
    <Route path="/bar" component={Bar} />,
    <Redirect from="/phoo" to="/foo" />,
    <Route path="/async" component={Foo} onEnter={onEnterWait} />,
    <Route path="/params(/:a)" component={Params} />,
  ]

  // Something to return 200 when we are not validating routes.
  const validPath = '/foo'

  function reducer() {
    return state
  }

  // Trivial koa middleware.
  function* addXPooHeader(next) {
    this.set('X-Poo', 'true')
    yield* next
  }

  // Start a single server for all tests.
  const server = new SmushServer({
    port,
    routes,
    reducer,
    mountPoint,
    initialState,
    scripts,
    styles,
    getTitle,
    middleware: [addXPooHeader],
  })
  before(function (done) {
    // This will throw if something else is already listening.
    server.run(done)
  })

  describe('.run(done)', function () {
    it('listens on the provided port', function (done) {
      get('/', (err) => {
        expect(err).is.null
        done()
      })
    })

    describe('uses the provided routes', function () {
      it('/ returns 404', function (done) {
        get('/', (err, resp) => {
          expect(resp.statusCode).to.be.equal(404)
          done()
        })
      })

      it('/foo returns 200 with <Foo />', function (done) {
        get('/foo', (err, resp, $) => {
          expect(resp.statusCode).to.be.equal(200)
          expectSelector($, 'body .foo', 1)
          expectSelector($, 'body .bar', 0)
          done()
        })
      })

      it('/bar returns 200 with <Bar />', function (done) {
        get('/bar', (err, resp, $) => {
          expect(resp.statusCode).to.be.equal(200)
          expectSelector($, 'body .bar', 1)
          expectSelector($, 'body .foo', 0)
          done()
        })
      })

      it('/phoo returns 302 to /foo', function (done) {
        get('/phoo', (err, resp, $) => {
          expect(resp.statusCode).to.be.equal(302)
          expect(resp.headers['location']).to.be.equal('/foo')
          done()
        })
      })

      it('/async returns 200 with <Foo /> after delay', function (done) {
        get('/async', (err, resp, $) => {
          expect(resp.statusCode).to.be.equal(200)
          expectSelector($, 'body .foo', 1)
          expectSelector($, 'body .bar', 0)
          done()
        })
      })
    })

    it('renders the app inside the mount point', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        // App should be only child of mount point.
        expectSelector($, `body > #${mountPoint} > *`, 1)
        // App should have a react doodad.
        expectSelector($, `body > #${mountPoint} > *[data-reactid]`, 1)
        done()
      })
    })

    it('serializes the initial state into an inline script tag', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        // Initial state should be only inline script.
        expectSelector($, 'body > script', 1)
        // Verify global variable name.
        expect($('body > script').text()).to.have.string(`window.${initialState} =`)
        done()
      })
    })

    it('uses the provided reducer', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        // Initial state is created by the reducer, so we verify against that
        expect($('body > script').text()).to.have.string(`= ${state}`)
        done()
      })
    })

    it('creates a body with just the app and initial state', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        expectSelector($, 'body > *', 2)
        done()
      })
    })

    it('inserts the provided scripts into head with defer', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        scripts.forEach(path => {
          expectSelector($, `head > script[src="${path}"][defer]`, 1)
        })
        done()
      })
    })

    it('inserts the provided styles into head', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        styles.forEach(path => {
          expectSelector($, `head > link[href="${path}"]`, 1)
        })
        done()
      })
    })

    it('calls the title provider to render the document title', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        expect($('title').text()).to.be.equal(getTitle())
        done()
      })
    })

    it('registers the provided koa middleware', function (done) {
      get(validPath, (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        // Request lowercases headers?
        expect(resp.headers['x-poo']).to.be.equal('true')
        done()
      })
    })

    it('supports params and query string in routes', function (done) {
      get("/params/1?b=2", (err, resp, $) => {
        expect(resp.statusCode).to.be.equal(200)

        expect($('div').attr('data-a')).to.be.equal('1')
        expect($('div').attr('data-b')).to.be.equal('2')
        done()
      })
    })
  })

  describe('.stop(done)', function () {
    it('closes the port', function (done) {
      server.stop(() => {
        // If server is stopped, nothing should be listening.
        get('/', (err) => {
          expect(err).is.not.null
          done()
        })
      })
    })
  })
})


src/common.js
import React from 'react'
import {Provider} from 'react-redux'
import {createStore, applyMiddleware} from 'redux'

const defaultRoutes = []

function defaultReducer(state = []) {
  return state
}

export default class Hut {
  constructor(opts = {}) {
    // The global variable that will hold the serialized redux store.
    this.initialState = opts.initialState || 'initialState'

    // The class of the DOM node where react will be mounted/rendered.
    this.mountPoint = opts.mountPoint || 'app'

    // The react-router routes.
    this.routes = opts.routes || defaultRoutes

    // The redux reducer.
    this.reducer = opts.reducer || defaultReducer

    // Array of redux middleware to be applied, in order.
    this.reduxMiddleware = opts.reduxMiddleware || []

    this._createStore = applyMiddleware.apply(null, this.reduxMiddleware)(createStore)
  }

  createStore(initialState) {
    return this.store = this._createStore(this.reducer, initialState)
  }

  createElement(Component, props) {
    const element = React.createElement(Component, props)
    return React.createElement(Provider, {store: this.store}, element)
  }
}


src/server.js
import koa from 'koa'
import {createServer} from 'http'
import React from 'react'
import {renderToString} from 'react-dom/server'
// TODO: react-router 1.1 renames to RouterContext
import {RoutingContext, match} from 'react-router'

import Hut from './common.js'

export default class HutServer extends Hut {
  constructor(opts = {}) {
    super(opts)

    // The port to bind on.
    this.port = opts.port || 3000

    // An ordered list of middleware to apply to koa.
    this.middleware = opts.middleware || []

    // The title used during server rendering. May also be a function.
    this.getTitle = opts.getTitle || (() => '')

    // Scripts to include unconditionally on every render.
    this.scripts = (opts.scripts || []).map(script => {
      return `<script type="text/javascript" src="${script}" defer></script>`
    }).join('')

    // Stylesheets to include unconditionally on every render.
    this.styles = (opts.styles || []).map(style => {
      return `<link rel="stylesheet" type="text/css" href="${style}">`
    }).join('')
  }

  run(cb) {
    const app = koa()

    this.middleware.forEach(mw => app.use(mw))

    // TODO: koa 2.0 passes context instead of binding to this.
    const self = this
    app.use(function* () {
      yield new Promise((resolve, reject) => self.renderSite(this, resolve))
    })

    this.server = createServer(app.callback())
    this.server.listen(this.port, cb)
  }

  stop(cb) {
    this.server.close(cb)
    this.server = null
  }

  renderSite(ctx, done) {
    match({routes: this.routes, location: ctx.url}, (err, redir, props) => {
      if (err) {
        ctx.status = 500
        ctx.body = `Error: ${err}`
        return done()
      }

      if (redir) {
        ctx.redirect(redir.pathname)
        ctx.status = 302
        ctx.body = `Moved: ${ctx.path} => ${redir.pathname}`
        return done()
      }

      if (!props) {
        ctx.status = 404
        ctx.body = `Bad route: ${ctx.path}`
        return done()
      }

      const store = this.createStore()

      // Populate the store before rendering.
      const promises = []
      props.routes.forEach(route => {
        (route.component.actions || []).forEach(action => {
          promises.push(store.dispatch(action(props)))
        })
      })

      Promise.all(promises).then(() => {
        const app = this.createElement(RoutingContext, props)
        const html = renderToString(app)
        const title = this.getTitle()
        const initialState = JSON.stringify(store.getState())

        ctx.set('Content-Type', 'text/html; charset=utf-8')
        ctx.body = `
        <!DOCTYPE html>
        <html>
        <head>
          <meta char-set="utf-8">
          <meta name="viewport" content="width=device-width, user-scalable=no">
          ${this.scripts}
          ${this.styles}
          <title>${title}</title>
        </head>
        <body>
          <section id="${this.mountPoint}">${html}</section>
          <script type="text/javascript">
            window.${this.initialState} = ${initialState}
          </script>
        </body>
        </html>
        `
        return done()
      }, (err) => {
        ctx.status = 500
        ctx.body = `Error: ${err}`
        return done()
      })
    })
  }
}
